Social Warning

Prashant-Android opened the pull request:
Added Social Warning in Specifc screeen .

Prashant-Android committed:
added social warning in MessageListScreen and ClazzAssignmentDetailOverviewScreen .

Prashant-Android committed:
added social warning strings in strings.xml

Prashant-Android committed:
removed hardcoded color value from SocialWarningListItem.kt

Anugraha-sutara commented on core/src/commonMain/resources/MR/base/strings.xml:
```diff
@@ -4,6 +4,12 @@
+    <string name="got_it">"Got it"</string>
+    <string name="learn_more">"Learn more"</string>
+    <string name="social_warning_one">"Avoid sharing personal information"</string>
+    <string name="social_warning_two">"Report any inappropriate behavior"</string>
```
Just a small suggestion  — instead of numbering the warnings as one, two, three, maybe we could try using descriptive names? It might make things a bit clearer 
`<string name="social_warning_no_personal_info">Avoid sharing personal information</string>
`
`<string name="social_warning_report_behavior">Report any inappropriate behavior</string>
`
`<string name="social_warning_respect_others">Be respectful in interactions</string>
`

Prashant-Android committed:
correct naming convention

Prashant-Android committed:
correct naming convention

Prashant-Android committed:
Fixed binding issue in jvmTest

Prashant-Android committed:
Adjust Ui.

mikedawson commented on app-react/src/jsMain/kotlin/com/ustadmobile/mui/components/SocialWarningListItem.kt:
```diff
@@ -0,0 +1,87 @@
+import com.ustadmobile.core.MR
+import com.ustadmobile.core.hooks.useStringProvider
+import mui.icons.material.Circle
+import mui.icons.material.Warning
```
As per note on app-react docs: import of icons must use the as - otherwise IDE performance is extremely poor. e.g. use
```
import mui.icons.material.Circle as CircleIcon
```

mikedawson commented on app-react/src/jsMain/kotlin/com/ustadmobile/mui/components/SocialWarningListItem.kt:
```diff
@@ -0,0 +1,87 @@
+
+    val strings = useStringProvider()
+
+    Paper {
```
As per coding style - use default styles. Please do not use the extra Paper - as per task description this should all be in a **single** ListItem.

mikedawson commented on app-react/src/jsMain/kotlin/com/ustadmobile/mui/components/SocialWarningListItem.kt:
```diff
@@ -0,0 +1,87 @@
+import web.cssom.px
+
+external interface SocialWarningProps : Props {
+    var cautions: List<String>
```
This list of cautions should be removed.

mikedawson commented:
![warning](https://github.com/user-attachments/assets/e429a721-9097-40ec-8e2d-5c3123a4b60e)

Warning list item headline text: "Be careful interacting online and in-app!"
Secondary text: "Be careful not to share any real world information unless you are completely sure who you are interacting with. Never agree to meet anyone you don't know in real life without your parent or guardian."

The buttons are part of the secondary text section of listitem itself.

Clicking learn more should open https://beinternetawesome.withgoogle.com/

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/clazzassignment/detailoverview/ClazzAssignmentDetailOverviewScreen.kt:
```diff
@@ -185,39 +187,32 @@ fun ClazzAssignmentDetailOverviewScreen(
-                    headlineContent = { Text(formattedDateTime)},
-                    supportingContent = { Text(stringResource(MR.strings.deadline)) }
-                )
+            if (uiState.cbDeadlineDateVisible) {
```
Revert what was not changed: This code section should not have been changed and now the indentation is wrong.

mikedawson reviewed (changes requested):
Please do not modify sections of code that have not been changed (autoformat or any other way). It makes doing the review much more complicated.

e.g. changing the spacing from:
```if(uiState.addFileSubmissionVisible) {```
to:
```if (uiState.addFileSubmissionVisible) {```

Please don't do that!

Please revert all such things so that they do not show in the changes. They should not be changed. Please self-check in the pull request changes to ensure that only what you intended to changed is showing in the changes.

mikedawson commented:
Also: please ensure the list of cautions concept (as a List<String>) is removed. There is a headline and subheadline (which should be added to strings) as above - that's going to be it.

Prashant-Android committed:
Merge remote-tracking branch 'origin/primary' into primary

Prashant-Android committed:
Merge branch 'refs/heads/primary' into dev-social-warning

Prashant-Android committed:
Reverted unintended formatting changes in unrelated files.
Removed unnecessary Paper and obsolete cautions list as per review feedback.
Updated SocialWarningListItem layout to match task requirements with headline, secondary text, and buttons below text.
Removed unnecessary Paper and obsolete cautions list as per review feedback.

Prashant-Android committed:
Fix String issue & social warning url.

Prashant-Android committed:
Fix String issue & social warning url.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/clazzassignment/detailoverview/ClazzAssignmentDetailOverviewViewModel.kt:
```diff
@@ -731,5 +756,7 @@ class ClazzAssignmentDetailOverviewViewModel(
 
         const val DEST_NAME = "CourseAssignmentDetailOverviewView"
 
+        const val SOCIAL_WARNING_WEB_URL = "https://beinternetawesome.withgoogle.com"
```
This constant should be declared on a related use case (e.g. ShowSocialWarningUseCase), not in this viewmodel. It should be declared once, such that if this link needs to change, there should be one place to go and change it.

This code as-is does implement single point of truth as per coding style - the same constant is being declared in multiple places.

mikedawson replied:
![Screenshot from 2024-11-26 12-42-43](https://github.com/user-attachments/assets/2dc36449-a7b5-44c8-9761-258c1ae295c7)

This is against the [coding style](https://github.com/UstadMobile/UstadMobile/blob/primary/CODING-STYLE.md#never-hardcode-any-literal-values) - "Any literal other than 0, 1, -1, true, false, or null. Any other literal value MUST be defined as a constant with a meaningful name."

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/components/SocialWarningListItem.kt:
```diff
@@ -0,0 +1,75 @@
+import com.ustadmobile.core.MR
+import dev.icerock.moko.resources.compose.stringResource
+
+@Composable
```
This must as per previous comment use the Jetpack Compose ListItem composable. That is what will provide consistent styles.

e.g. 
```
ListItem(
   headlineContent = { Text(MR.strings.be_careful_interacting_online) }, 
   supportingContent = {
       Column {
           Text(stringResource(MR.strings.be_careful_not_to_share)
           Row {
               //buttons can go here
           }
        }
   }
)
```

mikedawson replied:
![Screenshot from 2024-11-26 11-57-39](https://github.com/user-attachments/assets/fd458250-d407-4257-8d84-c429e2ac2fd0)

The revised version (as of this writing) is still clearly not consistent with the coding style:

As per https://github.com/UstadMobile/UstadMobile/blob/primary/CODING-STYLE.md#screens

"Screens must use default margins and colors provided by the platform theme (Jetpack Compose or MUI) unless specifically noted otherwise"

That means the text style, text color, and tint should NOT be manually set unless it is explicitly noted to do so.

mikedawson replied:
Additionally (as per [ListItem doc](https://developer.android.com/reference/kotlin/androidx/compose/material3/package-summary#ListItem(kotlin.Function0,androidx.compose.ui.Modifier,kotlin.Function0,kotlin.Function0,kotlin.Function0,kotlin.Function0,androidx.compose.material3.ListItemColors,androidx.compose.ui.unit.Dp,androidx.compose.ui.unit.Dp))) the icon should go in the leadingContent, you should not create a row within leadingContent or supportingContent.

mikedawson commented on app-react/src/jsMain/kotlin/com/ustadmobile/mui/components/SocialWarningListItem.kt:
```diff
@@ -0,0 +1,76 @@
+
+            ListItemText {
+                primary = ReactNode(strings[MR.strings.be_careful_interacting_online])
+                secondary = ReactNode(strings[MR.strings.be_careful_not_to_share])
```
The buttons are part of the ListItem and should be part of secondary text. The same is done in ClazzDetailOverviewCourseBlockListItem in app-react.

 ```
e.g.
ListItem {
    ListItemText {
      secondary = Stack.create {
         direction = responsive(StackDirection.column)
         + strings[MR.strings.text_constant]
       
         Stack {
            direction = responsive(StackDirection.row)
            //Buttons can go here   
          }
      }
   
     secondaryTypographyProps = jso {
        component = div
     }
   }  
}
```

Prashant-Android committed:
- Moved `SOCIAL_WARNING_WEB_URL` to `ShowSocialWarningUseCase`
- Buttons are part of secondary text with proper alignment in a responsive layout.

Prashant-Android committed:
Fixed ui.

mikedawson commented on app-react/src/jsMain/kotlin/com/ustadmobile/mui/components/SocialWarningListItem.kt:
```diff
@@ -0,0 +1,80 @@
+import web.window.window
+import mui.icons.material.Warning as WarningIcon
+
+private const val SOCIAL_WARNING_WEB_URL = "https://beinternetawesome.withgoogle.com"
```
As per previous comment https://github.com/UstadMobile/UstadMobile/pull/980#discussion_r1856554102
"

This constant should be declared on a related use case (e.g. ShowSocialWarningUseCase), not in this viewmodel. It should be declared once, such that if this link needs to change, there should be one place to go and change it.

This code as-is does implement single point of truth as per coding style - the same constant is being declared in multiple places.
"

Prashant-Android committed:
Fixed ui.
